spec/abstract/AARCH64/FPU_A.thy

```diff
@@ -0,0 +1,59 @@
+(*
+ * Copyright 2024, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+chapter "AARCH64 FPU Functions"
+
+theory FPU_A
+imports
+  KHeap_A
+begin
+
+context Arch begin arch_global_naming (A)
+
+definition save_fpu_state :: "obj_ref \<Rightarrow> (unit,'z::state_ext) s_monad" where
+  "save_fpu_state t \<equiv> do
+     fpu_state \<leftarrow> do_machine_op readFpuState;
+     as_user t (setFPUState fpu_state)
+   od"
+
+definition load_fpu_state :: "obj_ref \<Rightarrow> (unit,'z::state_ext) s_monad" where
+  "load_fpu_state t \<equiv> do
+     fpu_state \<leftarrow> as_user t getFPUState;
+     do_machine_op (writeFpuState fpu_state)
+   od"
+
+\<comment> \<open>FIXME: maybe use a case instead of the if (depends on if wpc or if_split is easier)\<close>
+definition switch_local_fpu_owner :: "obj_ref option \<Rightarrow> (unit,'z::state_ext) s_monad" where
+  "switch_local_fpu_owner new_owner \<equiv> do
+     do_machine_op enableFpu;
+     cur_fpu_owner \<leftarrow> gets (arm_current_fpu_owner \<circ> arch_state);
+     when (cur_fpu_owner \<noteq> None) $ save_fpu_state (the cur_fpu_owner);
```

Comment:
I would generally prefer `case` for option if you have both branches, because it allows you to avoid using `the`.

Reply:
That comment was actually for the `if` statement just after this `when`, which does have both branches. I'll have a dig around for a helpful combinator. Assuming one does exist, would we want something similar in the Haskell?

Reply:
Yes, having them structurally the same would be nice. (It's not that super important, it's correct as it is, just a style question, so if we don't already have a combinator, we can also leave it as is.)

Reply:
Context: the FIXME was originally because we weren't sure whether dealing with `if` or `case` would be easier in the `corres` proof, so the plan was to check and then remove the FIXME afterwards

Reply:
I've made this change now but it's a little bit more complicated in Haskell, as to the best of my knowledge there isn't an easily accessible equivalent to `maybeM`. The most likely option that I found is https://hackage.haskell.org/package/extra-1.7.16/docs/Control-Monad-Extra.html#v:whenJust, but using that would introduce a new dependency on the `extra` package. I've currently used the underlying `for` from `Data.Traversable`, but that feels a bit unreadable and will require some care when translating to the design spec. Thoughts?
https://github.com/seL4/l4v/blob/285ecfc1ee461291cdd239b28f6b7666203b2bc3/spec/haskell/src/SEL4/Object/FPU/AARCH64.hs#L33-L41

Reply:
I don't have a good solution here. There's a function `maybe` in `Data.Maybe` that we could use with `maybe (return ()) saveFpuState curFpuOwner` that we could make an alias for to remove the `return ()` part, but then the return values have to match so it would only work for unit returns. Otherwise you could `return undefined` with the correct type (whatever `'b` the function returns). How does `for` get through the Haskell translator? I don't think we have something like that in Isabelle.

Reply:
Yeah, `for` won't get through the translator without doing something manual, so on second thought probably shouldn't use it.

I think I'm ok with your `maybe` idea, particularly since now that I look at it, that's basically how `whenJust` is implemented. It lines up pretty well with `maybeM` as well, which requires a unit return.

Reply:
> Context: the FIXME was originally because we weren't sure whether dealing with `if` or `case` would be easier in the `corres` proof, so the plan was to check and then remove the FIXME afterwards

Good point, and also works as a reminder that there's two things going on here. There's switching the `when` to a `maybeM` which I think we all agree is nice to avoid `the`, and there's whether it should `if` or `case`, where `case` is again nice in that it avoids `the` but it is unknown which will be easier in the proofs. How about for now I switch it to `case` but leave the FIXME until we get to the proofs?

Reply:
yes, I'd be happy with that